sharding/node/backend.go

```diff
@@ -0,0 +1,187 @@
+		path = ctx.GlobalString(utils.DataDirFlag.Name)
+	}
+
+	endpoint := ctx.Args().First()
+	if endpoint == "" {
+		endpoint = fmt.Sprintf("%s/%s.ipc", path, mainchain.ClientIdentifier)
+	}
+	if ctx.GlobalIsSet(utils.IPCPathFlag.Name) {
+		endpoint = ctx.GlobalString(utils.IPCPathFlag.Name)
+	}
+
+	passwordFile := ctx.GlobalString(utils.PasswordFileFlag.Name)
+	depositFlag := ctx.GlobalBool(utils.DepositFlag.Name)
+	actorFlag := ctx.GlobalString(utils.ActorFlag.Name)
+
+	smcClient, err := mainchain.NewSMCClient(endpoint, path, depositFlag, passwordFile)
+	if err != nil {
+		return nil, err
+	}
+
+	// Adds the initialized SMCClient to the ShardEthereum instance.
+	shardEthereum.smcClient = smcClient
+
+	// TODO: Find a better way to do this???
+	if err := shardEthereum.registerP2P(); err != nil {
+		return nil, err
+	}
+
+	if err := shardEthereum.registerTXPool(); err != nil {
+		return nil, err
+	}
+
+	if err := shardEthereum.registerActorService(actorFlag); err != nil {
+		return nil, err
+	}
+
+	return shardEthereum, nil
+}
+
+// Start the ShardEthereum service and kicks off the p2p and actor's main loop.
+func (s *ShardEthereum) Start() error {
+	log.Println("Starting sharding node...")
+
+	services := make(map[reflect.Type]sharding.Service)
+	for _, constructor := range s.serviceFuncs {
+		// Create a new context for the particular service.
+		ctx := &sharding.ServiceContext{
+			Services: make(map[reflect.Type]sharding.Service),
+		}
+		// Copy needed for threaded access.
+		for kind, s := range services {
+			ctx.Services[kind] = s
+		}
+		// Construct and save the service.
+		service, err := constructor(ctx)
+		if err != nil {
+			return err
+		}
+		kind := reflect.TypeOf(service)
+		if _, exists := services[kind]; exists {
+			return fmt.Errorf("service already exists: %v", kind)
+		}
+		services[kind] = service
+	}
+
+	// Start each of the services.
+	started := []reflect.Type{}
+	for kind, service := range services {
+		// Start the next service, stopping all previous upon failure.
+		if err := service.Start(); err != nil {
+			for _, kind := range started {
+				services[kind].Stop()
+			}
+			return err
+		}
+		// Mark the service started for potential cleanup.
+		started = append(started, kind)
+	}
+	return nil
+}
+
+// Close handles graceful shutdown of the system.
+func (s *ShardEthereum) Close() error {
+	return nil
+}
+
+// SMCClient returns an instance of a client that communicates to a mainchain node via
+// RPC and provides helpful bindings to the Sharding Manager Contract.
+func (s *ShardEthereum) SMCClient() *mainchain.SMCClient {
+	return s.smcClient
+}
+
+// Register appends a service constructor function to the service registry of the
+// sharding node.
+func (s *ShardEthereum) Register(constructor sharding.ServiceConstructor) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	s.serviceFuncs = append(s.serviceFuncs, constructor)
+	return nil
+}
+
+func (s *ShardEthereum) registerP2P() error {
```

Comment:
We should probably add function comments for `registerP2P()` and `registerTXPool()` so the next person will know what to work on